### PR TITLE
Add pub mod option for UnlinkedFile

### LINKED ISSUE #8228: For unlinked files, add option to add "pub mod" instead of just "mod"
It seems to be a pretty common pattern to have mod.rs files that are just a bunch of `pub mod x`, `pub mod y` etc., so it would be nice when you create an unlinked file and use the code action from rust-analyzer, if in addition to `Insert 'mod x;',` it would have a `Insert 'pub mod x;'` option. Super useful feature already, just think this would make it even better :) 

Also you guys probably hear this a million times a day but rust-analyzer is amazing; it's totally changed the rust experience for me, and the pace of development is unreal.

### PR DESCRIPTION
**Pull request by rainy-me**
close #8228

This is a draft that changes `Diagnostic` to contain multiple fixes. Pre analysis is in https://github.com/rust-analyzer/rust-analyzer/issues/8228#issuecomment-812887085 Because this solution is straightforward so I decided to type it out for discussion.

Currently the `check_fix` is not able to test the situation when multiple fixes available. <del>Also because `Insert 'mod x;'` and  `Insert 'pub mod x;'` are so similar, I don't know how to test them correctly and want some suggestions.</del>. I added 
 `check_fixes` to allow checking mutiple possible fixes.

In additional, instead of append after possible existing `mod y`, I think it's possible to Insert `pub mod x;` after `pub mod y`. Should I implement this too?

### CODE REVIEW

File: crates/rust-analyzer/src/handlers.rs [hunk 1]
```diff
@@ -1033,7 +1033,8 @@ fn add_quick_fixes(
 
     for fix in diagnostics
         .into_iter()
-        .filter_map(|d| d.fix)
+        .filter_map(|d| d.fixes)
+        .flatten()
```
Comment (jonas-schievink) [inline crates/rust-analyzer/src/handlers.rs:1037]:
```suggestion
        .flat_map(|d| d.fixes)
```

Comment (rainy-me) [reply crates/rust-analyzer/src/handlers.rs:1037]: I thought about this but `d.fixes` here is an `Option<Vec<Fix>>` so flat_map doesn't seem to work.
One possible way is 
```rust
        .flat_map(|d| match d.fixes {
            Some(fixes) => fixes,
            _ => vec![],
        })
```
and the other is make `fixes` of `Diagnostic` a `Vec<Fix>` which maybe better but requires more changes to existing code to convert None to empty vec.

Comment (rainy-me) [reply crates/rust-analyzer/src/handlers.rs]: Changed to the the first one for now https://github.com/rust-analyzer/rust-analyzer/pull/8345/commits/6a8018cb01a577673f9cbc4ab48d91d892e92017

Comment (jonas-schievink) [reply crates/rust-analyzer/src/handlers.rs]: Ah, I see, that seems good too

File: crates/ide/src/diagnostics/fixes.rs [hunk 2]
```diff
@@ -22,18 +22,18 @@ use text_edit::TextEdit;
...
 use crate::{diagnostics::Fix, references::rename::rename_with_semantics, FilePosition};
 
-/// A [Diagnostic] that potentially has a fix available.
+/// A [Diagnostic] that potentially has some fixes available.
 ///
 /// [Diagnostic]: hir::diagnostics::Diagnostic
-pub(crate) trait DiagnosticWithFix: Diagnostic {
-    fn fix(&self, sema: &Semantics<RootDatabase>) -> Option<Fix>;
+pub(crate) trait DiagnosticWithFixes: Diagnostic {
+    fn fixes(&self, sema: &Semantics<RootDatabase>) -> Option<Vec<Fix>>;
```
Comment (jonas-schievink) [inline crates/ide/src/diagnostics/fixes.rs:29]: Is there an important invariant that the vector must be non-empty if this returns `Some`? If not I think this should just return `Vec<Fix>` without the option.

Comment (rainy-me) [reply crates/ide/src/diagnostics/fixes.rs]: As I explained in https://github.com/rust-analyzer/rust-analyzer/pull/8345#discussion_r607323890 as a second option, I think `Vec<Fix>` is better too. However functions inside `fix` currently make extensive use of `Try` trait and turning them into `Vec<Fix>` without `Option` doesn't seem easy to do.

e.g:
https://github.com/rust-analyzer/rust-analyzer/blob/10a243ea55565a0dd1de52f8f802c3e3a7bfef54/crates/ide/src/diagnostics/fixes.rs#L162-L176

Comment (rainy-me) [reply crates/ide/src/diagnostics/fixes.rs]: Alternatively, I think it also makes sense to change it to return something like `Option<HashMap<AssistId,Assist>>` which would possibly make it easier to test a specific Assist.

**Review by jonas-schievink — APPROVED** [review]
This seems good to go, thanks!

bors d+

**Comment by rainy-me** [comment]
I resolved recent conflicts but think it's better to run the github workflow before r+?

**Pull request merged**